server/supernova/sc/sc_osc_handler.cpp

```diff
@@ -2166,13 +2166,23 @@ template <bool realtime> void handle_b_write(ReceivedMessage const& msg, endpoin
             try {
                 sc_factory->buffer_write(bufnum, filenameString.c_str(), headerString.c_str(), sampleString.c_str(),
                                          start, frames, leave_open);
-                message.trigger_async(endpoint);
-                cmd_dispatcher<realtime>::fire_done_message(endpoint, b_write, bufnum);
+
+                cmd_dispatcher<realtime>::fire_rt_callback(
+                    [=, message = std::move(message), filenameString = std::move(filenameString),
+                     headerString = std::move(headerString), sampleString = std::move(sampleString)]() mutable {
+                        handle_completion_message(std::move(message), endpoint);
+
+                        consume(std::move(filenameString));
+                        consume(std::move(headerString));
+                        consume(std::move(sampleString));
+
+                        cmd_dispatcher<realtime>::fire_done_message(endpoint, b_write, bufnum);
+                    });
             } catch (std::exception const& error) {
                 report_failure(endpoint, error, b_write, bufnum);
+                cmd_dispatcher<realtime>::free_in_rt_thread(std::move(message), std::move(filenameString),
+                                                            std::move(headerString), std::move(sampleString));
```

Comment:
@sonoro1234 have you tested this code path? from what I can see, if `fire_rt_callback` throws and we end up here, all these strings will have been moved-from, and therefore empty. i'm not sure if there's a clear path where that lambda will throw; if you need to you could force this path by manually inserting a `throw std::runtime_error("");` in the middle of the `fire_rt_callback` lambda.

Reply:
I'm using the same pattern here as in `handle_b_alloc`, `handle_b_allocRead`, etc. 

Generally, `fire_rt_callback` shouldn't throw. Even if it *would* throw, I don't see a problem because a moved from `movable_string` is in a well-defined state (`data_ = nullptr`).

Reply:
> Even if it would throw, I don't see a problem because a moved from movable_string is in a well-defined state (data_ = nullptr).

OK, I see now that `free_in_rt_thread` doesn't do anything with its arguments except sink them. A comment here to that effect would be pretty helpful, since it's nonobvious at first that this isn't a programmer error (move semantics are tricky to reason about), and I had to jump around a bit to see it.

Reply:
What I have just tried is b_write with an inexistent path. There is no /fail message althought the files are not saved.
This is because `sc_factory->buffer_write` doen't throw but instead return 0 on succes and -1 on failure.
https://github.com/supercollider/supercollider/blob/develop/server/supernova/sc/sc_plugin_interface.cpp#L937

In fact I have not been able to track something throwing in this code path.

@Spacechild1 `buffer_write`  failure should be processed somehow according to return value.

Reply:
> A comment here to that effect would be pretty helpful, since it's nonobvious at first that this isn't a programmer error

@brianlheim 
As @Spacechild1 says, it is the same happening in other codes as  handle_b_alloc, handle_b_allocRead, etc. which have no comment about it.

Reply:
not sure how much it matters that this follows a pattern, the pattern is still confusing and IMO deserves a comment _somewhere_.

Reply:
I could add a comment to the `free_in_rt_thread` function, so users just have to jump to the function definition to see immediately what's going on.

Reply:
never mind about the comment, i don't want to make a fuss about it.

Reply:
@brianlheim so please approve or insist in asking for the comment.